netket/optimizer/sr/s_jacobian_mat.py

```diff
@@ -0,0 +1,218 @@
+    mode: str,
+    rescale_shift: bool,
+):
+    """Calculates the gradients O_ij by backpropagating every sample separately,
+    vectorising the loop using vmap
+    If rescale_shift is True, columns of O are rescaled to unit norm, and
+    scale factor [1/sqrt(S_kk)] returned as a separate vector for
+    scale-invariant regularisation as per Becca & Sorella p. 143.
+    """
+    # Ravel the parameter PyTree and obtain the unravelling function
+    params, unravel = nkjax.tree_ravel(params)
+
+    if jnp.ndim(samples) != 2:
+        samples = jnp.reshape(samples, (-1, samples.shape[-1]))
+    n_samples = samples.shape[0]
+
+    if mode == "holomorphic":
+        # Preapply the model state so that when computing gradient
+        # we only get gradient of parameters
+        # Also divide through sqrt(n_samples) to normalise S matrix in the end
+        def fun(W, σ):
+            return (
+                apply_fun({"params": unravel(W), **model_state}, σ[jnp.newaxis, :])[0]
+                / n_samples ** 0.5
+            )
+
+        grads = _grad_vmap_minus_mean(fun, params, samples, True)
+    elif mode == "R2R":
+
+        def fun(W, σ):
+            return (
+                apply_fun({"params": unravel(W), **model_state}, σ[jnp.newaxis, :])[0].real
+                / n_samples ** 0.5
+            )
+
+        grads = _grad_vmap_minus_mean(fun, params, samples, False)
+    elif mode == "R2C":
+
+        def fun1(W, σ):
+            return (
+                apply_fun({"params": unravel(W), **model_state}, σ[jnp.newaxis, :])[0].real
+                / n_samples ** 0.5
+            )
+
+        def fun2(W, σ):
+            return (
+                apply_fun({"params": unravel(W), **model_state}, σ[jnp.newaxis, :])[0].imag
+                / n_samples ** 0.5
+            )
+
+        # Stack real and imaginary parts as real matrixes along the "sample"
+        # axis to get Re(O†O) directly
+        grads = jnp.concatenate(
+            (
+                _grad_vmap_minus_mean(fun1, params, samples, False),
+                _grad_vmap_minus_mean(fun2, params, samples, False),
+            ),
+            axis=0,
+        )
+    else:
+        raise Exception(
+            "Differentation mode must be holomorphic, R2R or R2C, got {}".format(mode)
+        )
+
+    if rescale_shift:
+        sqrt_Skk = jnp.linalg.norm(grads, axis=0, keepdims=True)
+        return grads / sqrt_Skk, jnp.ravel(sqrt_Skk)
+    else:
+        return grads
+
+def _grad_vmap_minus_mean(
+    fun: Callable, params: jnp.ndarray, samples: jnp.ndarray, holomorphic: bool
+):
+    """Calculates the gradient of a neural network for a number of samples
+    efficiently using vmap(grad), 
+    and subtracts their mean for each parameter, i.e., each column
+    """
+    grads = jax.vmap(
+        jax.grad(fun, holomorphic=holomorphic), in_axes=(None, 0), out_axes=0
+    )(params, samples)
+    return grads - jnp.sum(grads, axis=0, keepdims=True) / grads.shape[0]
```

Comment:
```suggestion
    return grads - sum_inplace(jnp.sum(grads, axis=0, keepdims=True)) / grads.shape[0]
```

Reply:
actually, you sum_inplace returns also a token object (second return value) that you should trash

Reply:
I think it's taken care of [here](https://github.com/attila-i-szabo/netket/blob/8d9909e311be8d7860fc9661cce1eb3efd32b34c/netket/stats/_sum_inplace.py#L99)

Reply:
ah yes